tls/s2n_handshake_io.c

```diff
@@ -823,19 +825,22 @@ static int handshake_read_io(struct s2n_connection *conn)
      */
     S2N_ERROR_IF(record_type == TLS_APPLICATION_DATA, S2N_ERR_BAD_MESSAGE);
     if (record_type == TLS_CHANGE_CIPHER_SPEC) {
+        /* TLS1.2 should not receive unexpected change cipher spec messages, but TLS1.3 might. */
+        S2N_ERROR_IF(!IS_TLS13_HANDSHAKE(conn) && EXPECTED_RECORD_TYPE(conn) != TLS_CHANGE_CIPHER_SPEC, S2N_ERR_BAD_MESSAGE);
+
         S2N_ERROR_IF(s2n_stuffer_data_available(&conn->in) != 1, S2N_ERR_BAD_MESSAGE);
 
         GUARD(s2n_stuffer_copy(&conn->in, &conn->handshake.io, s2n_stuffer_data_available(&conn->in)));
-        GUARD(ACTIVE_STATE(conn).handler[conn->mode] (conn));
+        GUARD(CCS_STATE(conn).handler[conn->mode] (conn));
```

Comment:
This seems to override the state machine's active state handler with CCS's handler. What happens after the CCS handler is called, can we be sure the active state's handler still gets called correctly? At the end of this block, should the code break instead of returning 0?

Reply:
Think of an unexpected CCS message in TLS1.3 as a no-op for the state machine. It's processed, but doesn't affect the actual state of the handshake. We don't advance the state machine. So the active state is still the active state, and its handler will be called when the correct handshake message for that state is received. The existing unit tests verify that we don't advance the state machine when we receive an unexpected CCS message.

Returning 0 is correct-- we're done processing this record. A CCS record can't contain any other record types, like handshake messages.